### PR TITLE
Updated Sig-windows Memory Limits tests to not assume all nodes are the same

### LINKED ISSUE #106608: Update Sig-windows Memory Limits tests to not assume all nodes are the same
### What happened?

The test `[sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] attempt to deploy past allocatable memory limits should fail deployments of pods once there isn't enough memory`: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-aks-engine-azure-master-staging-windows-serial-slow/1461874338100678656 is failing because it assumes the memory is the same across all nodes.

A change in the configuration https://github.com/kubernetes-sigs/windows-testing/pull/293 changed that assumption and the test started to fail.

### What did you expect to happen?

the test to assign the correct amount of memory to each of the nodes to fill them up.

### How can we reproduce it (as minimally and precisely as possible)?

create a Windows cluster with 3 nodes.  Two have 6gb of allocatable memory, 1 has 14gb of allocatable memory.  Run the test above and it will fail because it tries to create 3 pods with 9 gb of memory.  The first pod will land on the node with 14gb of memory but the other two pods won't be able to find a node with enough memory (9gb pod > 6 gb nodes).

### Anything else we need to know?

https://github.com/kubernetes-sigs/windows-testing/pull/297 fixes this in the configuration

/sig windows

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
#### What type of PR is this?
Fixing Bug 

/kind bug
#### What this PR does / why we need it:
Updated Sig-windows Memory Limits tests to not assume all nodes are the same. 
#### Which issue(s) this PR fixes:

Fixes #106608

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
@jsturtevant Am I understanding correct here :
- We have to add `pods = newMemLimitTestPods(1, imageutils.GetPauseImageName(), podType, getTotalAllocatableMemory(f))` in the loop.
- And `failurePods = newMemLimitTestPods(1, imageutils.GetPauseImageName(), podType, getTotalAllocatableMemory(f))` outside the loop.

Do we need to use node selector in  func `newMemLimitTestPods()` as an argument because you mentioned it in your [review](https://github.com/kubernetes/kubernetes/pull/106985#discussion_r781627226), if yes where and how can we get correct one, I cannot find where we need it.

File: test/e2e/windows/memory_limits.go
Comment (NikhilSharmaWe) [inline test/e2e/windows/memory_limits.go:126]: @jsturtevant Is this is right method for adding image name, node selector in Pod spec. If yes, Could you please tell about how to find the correct memory to be assigned to each pod.

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:127]: Yes this is how you would create a pod. You will want to capture the result and to verify the pod started:
https://github.com/kubernetes/kubernetes/blob/7dd03c8326858ceb232d38476db3240a15e8494f/test/e2e/windows/dns.go#L61-L62

Add memory limit like this: https://github.com/kubernetes/kubernetes/blob/7dd03c8326858ceb232d38476db3240a15e8494f/test/e2e/windows/cpu_limits.go#L127-L132

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:127]: you can get that by querying the node for allocatable memory see an example in the `getTotalAllocatableMemory` function

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:127]: @jsturtevant 

- I think we have to add only  `v1.ResourceMemory: status.Allocatable[v1.ResourceMemory]` in `Limits: v1.ResourceList`. 

Am I thinking correct here, do we also have to value for add `v1.ResourceCPU`.

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:127]: yes we only need memory in this case

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:127]: @jsturtevant in case of `pod` we can add `v1.ResourceMemory: status.Allocatable[v1.ResourceMemory]` for each `node.status`, but what should be added as memory for `failurePod`.

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:127]: i can be anything > than what is left over in the total Allocatable in the cluster for windows, since we have used it all up.  something like 500MB or 1GB should be good.

File: test/e2e/windows/memory_limits.go
Comment (NikhilSharmaWe) [inline test/e2e/windows/memory_limits.go:132]: @jsturtevant I think `&pod` should be correct argument for this case and `&failurePod` for the next case but it is showing an error. It is saying that pointer type should not be used but in [this](https://github.com/kubernetes/kubernetes/blob/7dd03c8326858ceb232d38476db3240a15e8494f/test/e2e/windows/dns.go#L61-L62) example pointer to `v1.Pod` type is used in the argument.

What is the cause of the error here ?

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:132]: Tried with only `pod` and `failurePod` too, but still it is showing an error.

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:132]: @jsturtevant Could you please give your thoughts on the above comment ?

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:132]: @jsturtevant Could you please give your thoughts on the above comment ?

**Comment by NikhilSharmaWe** [comment]
@jsturtevant Waiting for your response on [this](https://github.com/kubernetes/kubernetes/pull/107477#discussion_r784513109) comment and further code review.

**Comment by NikhilSharmaWe** [comment]
@jsturtevant Could you please respond to [this](https://github.com/kubernetes/kubernetes/pull/107477#discussion_r784513109) comment to move forward in the PR.

File: test/e2e/windows/memory_limits.go
Comment (jsturtevant) [inline test/e2e/windows/memory_limits.go:115]:
```suggestion
		pod := &v1.Pod{
```

File: test/e2e/windows/memory_limits.go
Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:132]:
```suggestion
		pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(context.TODO(), pod, metav1.CreateOptions{})
```

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:132]: This with the change https://github.com/kubernetes/kubernetes/pull/107477/files#r798791328 should work

File: test/e2e/windows/memory_limits.go
Comment (jsturtevant) [inline test/e2e/windows/memory_limits.go:115]: looking at the recent test failure the `newMemLimitTestPods` and `getTotalAllocatableMemory` are no longer used and need to be deleted.

Comment (NikhilSharmaWe) [reply test/e2e/windows/memory_limits.go:115]: @jsturtevant Yes true,

Could you please help me with this error
-  test/e2e/windows/memory_limits.go:132:3: `ineffectual assignment to pod (ineffassign)`.

Comment (jsturtevant) [reply test/e2e/windows/memory_limits.go:115]: from https://github.com/gordonklaus/ineffassign

>An assignment is ineffectual if the variable assigned is not thereafter used.

you are assigning pod but it's not used again. you can use `_` instead as an example.

**Comment by NikhilSharmaWe** [comment]
@jsturtevant I think PR is ready to merge, please inform if it needs improvements.

File: test/e2e/windows/memory_limits.go
Comment (jsturtevant) [inline test/e2e/windows/memory_limits.go:132]: Both this and the other pod spec are missing the pod name and container name (see the original pod spec definition for reference)

I would suggest this one be called something like "mem-failure-pod" and the other pods be the "mem-test-<nodename>"

**Review by jsturtevant — CHANGES_REQUESTED** [review]

File: test/e2e/windows/memory_limits.go
Comment (jsturtevant) [inline test/e2e/windows/memory_limits.go:112]: this pod spec should also include a `node: node.Name` field to make it deploy to that node

**Comment by jsturtevant** [comment]
Thanks for sticking with this one!  

The tests passed on a cluster with different size VMs:

```
k get nodes -l "kubernetes.io/os=windows" -o json | jq '.items[].status.allocatable.memory'

"8285748Ki"
"8285748Ki"
"16674356Ki"
"16674356Ki"
```

```
Mar 11 08:52:31.809: INFO: Found FailedScheduling event with message 0/5 nodes are available: 1 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 4 Insufficient memory. preemption: 0/5 nodes are available: 1 No victims found on node capz-conf-4pkcs for preemptor pod mem-failure-pod, 1 No victims found on node capz-conf-6q4qt for preemptor pod mem-failure-pod, 1 No victims found on node capz-conf-cmqhw for preemptor pod mem-failure-pod, 1 No victims found on node capz-conf-l5nsx for preemptor pod mem-failure-pod, 1 Preemption is not helpful for scheduling.
[AfterEach] [sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow]
  test/e2e/framework/framework.go:186
Mar 11 08:52:31.809: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "memory-limit-test-windows-8971" for this suite.
•{"msg":"PASSED [sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] attempt to deploy past allocatable memory limits should fail deployments of pods once there isn't enough memory","total":1,"completed":1,"skipped":4182,"failed":0}
```

**Comment by jsturtevant** [comment]
unrelated flakes
/retest

**Pull request merged**
